feat: introduce UA Blocker Middleware

finxol opened the pull request:
Create a new middleware for AI bots blocking.
Inspired by [ai.robots.txt](https://github.com/ai-robots-txt/ai.robots.txt).

This provides several options:
- Block all known AI bots
- Block all bots that we're not sure respect `robots.txt`
- Create a `robots.txt` file to be served (also offered as a middleware)

See README for details 

### The author should do the following, if applicable

- [x] Add tests
- [x] Run tests
- [x] `yarn changeset` at the top of this repo and push the changeset
- [x] Follow [the contribution guide](https://github.com/honojs/middleware?tab=readme-ov-file#how-to-contribute)

finxol committed:
feat: create AI bot blocker middleware

finxol committed:
chore: add changeset

finxol committed:
fix: uppercase function called too late

finxol committed:
chore: don't version-control robots.json

yusukebe commented on packages/ai-robots-txt/package.json:
```diff
@@ -0,0 +1,54 @@
+{
+  "name": "@hono/ai-robots-txt",
+  "version": "0.1.0",
```
The Changeset will bump the version number automatically. So this number should be `0.0.0`.

yusukebe commented on .changeset/early-grapes-walk.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@hono/ai-robots-txt': major
```
`major` is also fine, but I think `minor` is better. It may be early to make it as version `1.x`.

yusukebe commented on packages/ai-robots-txt/.gitignore:
```diff
@@ -0,0 +1 @@
+robots.json
```
It's difficult to decide whether to include `robots.json` or not. In my opinion, it's better to include `robots.json` in this repo. 

If you want to update `robots.json`, you will have to build and release a new version anyway (right?). It is not that much work to include `robots.json` in this project. Instead, versioning `robots.json` keeps track of the history, and allows users to track changes in `CHANGELOG.md` as well.

finxol replied:
Good points, this wasn't a hard decision, I'll remove it from gitignore.
It'll also allow any manual changes that might be needed/wanted in the future.

With this, we can automatically open a PR here whenever `robots.json` changes in `ai-robots-txt/ai.robots.txt` to update it here and flag the need for a new release.

finxol replied:
I've added a workflow to run the fetch script and open a PR to merge it if it has changed

yusukebe replied:
@finxol Thanks.

But, there is another idea. Can you check [the comment](https://github.com/honojs/middleware/pull/1181#discussion_r2134926296)?

yusukebe commented on packages/ai-robots-txt/package.json:
```diff
@@ -0,0 +1,54 @@
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/honojs/middleware.git",
+    "directory": "packages/hello"
```
This should be `packages/ai-robots-txt`.

yusukebe commented:
Can you add a CI for this middleware, referring to https://github.com/honojs/middleware/blob/main/.github/workflows/ci-hello.yml?

finxol committed:
chore: track `robots.json`

finxol committed:
ci: add `@hono/ai-robots-txt` workflow script

finxol committed:
fix: change initial version

finxol committed:
feat: add automatic `robots.json` sync from upstream

BarryThePenguin commented on packages/ai-robots-txt/package.json:
```diff
@@ -0,0 +1,54 @@
+    "publint": "attw --pack && publint",
+    "typecheck": "tsc -b tsconfig.json && echo 'Type checking completed'",
+    "test": "vitest",
+    "getrobotstxt": "node ./src/get-robots-txt.js"
```
This package could be published with an executable ([`bin`](https://docs.npmjs.com/cli/v11/configuring-npm/package-json#bin)) that consumers call to generate their own `robots.json` file. The output would then be passed as an argument to `blockAiBots()`

That way this repository doesn't need to be concerned with keeping the contents of the file up to date. It just determines the contents of the file

finxol replied:
@BarryThePenguin that’s a good idea, but then it would just be a user-agent based blocker.
Maybe we can export the generic as you suggest, but also export separately the `robots.json` file, which can then be passed to the generic blocker for easy setup?

BarryThePenguin replied:
I'm not too worried either way. I guess it's a trade off in convenience for the package consumer vs the package maintainer. Using the scheduled workflow requires someone to review/approve/publish any changes. If the format of the robots.json file changes, that requires the maintainer to handle those changes. However, keeping those changes internal to the package is a plus for consumers as it means the API remains small.

BarryThePenguin commented on packages/ai-robots-txt/src/index.ts:
```diff
@@ -0,0 +1,65 @@
+  return `(${formatted})`
+}
+
+export function blockAiBots(params = { allowRespecting: false }) {
+  return createMiddleware(async (c, next) => {
+    const userAgent = c.req.header('User-Agent')?.toUpperCase()
```
This could be taken a step further by having the `getrobotstxt` script generate two files, one with a compiled regex using the `respect !== 'Yes'` condition, and the other with a complied regex using all the entries.

Having a static regex at build time would remove the cost of constructing the regex

finxol committed:
feat!: change package name and architecture

finxol commented:
I've addressed all comments in the latest commit, the architecture and goal of the package is now broader than just an AI bots blocker.

The main export is now a generic User Agent based blocker, with a path export for the `robots.json` stuff that can be passed to the generic.
The regex construction is moved out of the middleware, to the `uaBlocker` function. It isn't done at build time, but enables the consumer to pass any list of user agents to block.

BarryThePenguin commented:
> `publint` fails for `@hono/ua-blocker/ai-bots` on node10 because of the export path. Do we need to support this `node10` module resolution, or can we ignore it?

Yes, it can be ignored by adding `attw --pack --profile node16` to the `publint` script.

I included this change, along with some other suggestions, in finxol#1. Feel free to include them or not 😅

BarryThePenguin reviewed (approved):
Your Readme and tests are very comprehensive 👏🏻

yusukebe commented:
@finxol 

Wooow. Making this UA Blocker Middleware for general user agents is a great idea. Including `ai-bots` is also a good idea. If the CI passes, I'll review again, merge it, and release!

finxol commented:
@yusukebe Thank you for the feedback! I just want to make a small change first, I'll push it later today and I'll let you know. 

And thank you for the great work on Hono :)

BarryThePenguin committed:
refactor(ua-blocker): prebuild compiled regex (#1)

finxol committed:
fix: add json data files to tsconfig

finxol committed:
chore: rename workflow files

finxol committed:
fix: test if string _contains_ "Yes"

It might be a markdown link, so not the exact string, but "[Yes](<link>)"

finxol committed:
fix: tests reflect the fixed "Yes" check

finxol committed:
feat: move generator back to prebuild

generated regex should use the version-controled robots.json, not
directly the upstream file

finxol committed:
chore: add .zed

finxol committed:
chore: remove unused files

finxol committed:
fix: properly setup workspace before running scripts

finxol committed:
chore: remove `prebuild` script from `build`, `typecheck`, and `test`

finxol committed:
chore: run `getrobotstxt` and `prebuild`

finxol committed:
fix: export `RegExp`s, not `string[]`s

finxol committed:
chore: mention RegExp and uppercase matching in docs

yusukebe commented:
@finxol

I tried using this UA Blocker middleware, and it felt good. The rest is making CI/test pass.

finxol committed:
fix: adapt tests to regex exports

finxol committed:
chore: add tests for direct regex passing

finxol committed:
chore: format code

yusukebe reviewed (approved):
LGTM!

yusukebe commented:
@finxol 

Thanks! Let's go with this. I'll merge and release an initial version.
